<a href="https://colab.research.google.com/github/saisumanth1902/L-square-QTify/blob/main/Session_4_OpenAI_API.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install -q openai

## OpenAI API Basics

Let's create a new OpenAI client and use it to generate a response to the user's question.

In [ ]:
from openai import OpenAI

client = OpenAI(api_key='Some Random value')

response = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[
        {"role": "user", "content": "Hello! How are you?"}
    ]
)

print(response)

AuthenticationError: Error code: 401 - {'error': {'message': 'Incorrect API key provided: Some Ran*****alue. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}

## OpenAI API Key

This is going to load the API key from the .env file.

In [ ]:
import os
from google.colab import userdata

OPENAI_API_KEY = userdata.get('OPENAI_API_KEY')
os.environ["OPENAI_API_KEY"] = OPENAI_API_KEY
os.environ["OPENAI_API_KEY_ERROR"] = "Wrong API Key"

## OpenAI API Response

Let's retry the same code, but by passing the API key as a parameter.

In [ ]:
from openai import OpenAI

client = OpenAI(api_key=OPENAI_API_KEY)

response = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[
        {"role": "user", "content": "Hello! How are you?"}
    ]
)

print(response)

ChatCompletion(id='chatcmpl-BWljTl6UcPwny6FmwRxjObCh9B3d6', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content="Hello! I'm just a chatbot so I don't have feelings, but I'm here to help you. How can I assist you today?", refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))], created=1747149679, model='gpt-3.5-turbo-0125', object='chat.completion', service_tier='default', system_fingerprint=None, usage=CompletionUsage(completion_tokens=31, prompt_tokens=13, total_tokens=44, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=0, audio_tokens=0, reasoning_tokens=0, rejected_prediction_tokens=0), prompt_tokens_details=PromptTokensDetails(audio_tokens=0, cached_tokens=0)))


## Response

Printing only the response content.

In [ ]:
print(response.choices[0].message.content)

Hello! I'm just a chatbot so I don't have feelings, but I'm here to help you. How can I assist you today?


## Different Models

Let's try to generate an image using GPT. You can click on the image link to see the image.

In [ ]:
response = client.images.generate(
   model="dall-e-2",
   prompt="Traffic with multiple supercars",
   n=1,
   size="512x512"
)

print(response)

ImagesResponse(created=1747150130, data=[Image(b64_json=None, revised_prompt=None, url='https://oaidalleapiprodscus.blob.core.windows.net/private/org-vezYUtUuPSSYHgIxYnniZxE4/user-auf1V9TNHptTVhs0XzyQi2Iq/img-votzvgHAuFIrTZl7sj72l7ru.png?st=2025-05-13T14%3A28%3A50Z&se=2025-05-13T16%3A28%3A50Z&sp=r&sv=2024-08-04&sr=b&rscd=inline&rsct=image/png&skoid=cc612491-d948-4d2e-9821-2683df3719f5&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2025-05-12T23%3A43%3A55Z&ske=2025-05-13T23%3A43%3A55Z&sks=b&skv=2024-08-04&sig=tU1eD32hWdH7EmCqoqGOagqh95K/Re%2BOXFQJLA7/OdM%3D')], usage=None)


## Temperature

### Low Temperature (0.0)

In [ ]:
response = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[
        {"role": "user", "content": "List all the items that you would include on a my table."}
    ],
    temperature=0.0
)
print(response.choices[0].message.content)

1. Tablecloth
2. Plates
3. Silverware
4. Glasses
5. Napkins
6. Centerpiece
7. Salt and pepper shakers
8. Candles
9. Table runner
10. Place mats


### High Temperature (1.0)

In [ ]:
response = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[
        {"role": "user", "content": "List all the items that you would include on a my table."}
    ],
    temperature=1.0
)
print(response.choices[0].message.content)

1. Tablecloth 
2. Plates 
3. Napkins 
4. Silverware 
5. Glasses 
6. Centerpiece 
7. Salt and pepper shakers 
8. Candles 
9. Serving platters 
10. Place cards 
11. Table runner 
12. Coasters 
13. Wine bottle opener 
14. Bread basket 
15. Dessert plates 
16. Table decorations 
17. Drinkware (such as wine glasses, water glasses, and coffee mugs) 
18. Table mats or placemats 
19. Condiment dishes 
20. Cloth napkins


## max_tokens

### 50 tokens

In [ ]:
response = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[
        {"role": "user", "content": "List all the items that you would include on a my table."}
    ],
    max_tokens=50
)
print(response.choices[0].message.content)

Tablecloth, plates, utensils, glasses, napkins, centerpiece, salt and pepper shakers, water pitcher, serving platters and bowls, bread basket, candles, place cards, and a table runner.


### 200 tokens

In [ ]:
response = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[
        {"role": "user", "content": "List all the items that you would include on a my table."}
    ],
    max_tokens=200
)
print(response.choices[0].message.content)

1. Tablecloth
2. Dinner plates
3. Salad plates
4. Napkins
5. Silverware (forks, knives, spoons)
6. Drinking glasses
7. Water pitcher
8. Centerpiece
9. Salt and pepper shakers
10. Candlesticks
11. Bread basket
12. Butter dish
13. Serving platters and bowls
14. Trivet for hot dishes
15. Place cards for seating arrangements


## Structured Outputs

Let's use prompt engineering to get a structured output.

In [ ]:
response = client.chat.completions.create(
    model="gpt-4o",
    messages=[
        {
            "role": "user",
            "content": [
                {
                    "type": "text",
                    "text": """You are an AI assistant for an e-commerce platform that automatically generates structured product details.
                    Given a product image, return the details in the following JSON format without additional text or commentary:
                    {
                        "name": "<Concise and accurate product title>",
                        "description": "<Detailed product description>",
                        "color": "<Primary color(s) of the product>",
                        "material": "<Material of the product>"
                    }"""
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": "https://www.irth.in/dw/image/v2/BKMH_PRD/on/demandware.static/-/Sites-Irth-product-catalog/default/dw746ec8b8/images/Irth/hi-res/R24CHHSML01PL1_1.jpg?sw=640"
                    }
                }
            ]
        }
    ]
)

json_response = response.choices[0].message.content

print(json_response)

```json
{
    "name": "Elegant Leather Handbag",
    "description": "A sophisticated handbag with a smooth, pebbled texture and a sleek design, featuring a sturdy handle and gold-tone accents.",
    "color": "Dark Purple",
    "material": "Leather"
}
```


## Let's build a workflow

The workflow consists of the following steps:

1. Create a story (text generation)
2. Visualize a scene (image generation)
3. Narrate the story (text-to-speech conversion)

Let's start with the first step.

We will use GPT-4o to generate a story.

In [ ]:
response = client.chat.completions.create(
    model="gpt-4o",
    messages=[
        {"role": "user", "content": "A story about a young boy who discovers he has the ability to fly in under 100 words."}
    ]
)

story = response.choices[0].message.content
print(story)

In a small, quiet village, young Sam discovered his incredible gift one breezy afternoon. Playing near a tall oak tree, he stumbled and instinctively flapped his arms. To his amazement, he lifted off the ground, soaring above fields of golden wheat. Initially frightened but quickly exhilarated, Sam experimented, zipping between clouds, feeling the freedom of the skies. Returning to the ground, he kept his gift secret, savoring private flights at dusk. With each ascent, Sam dreamed of future adventures, knowing his life was destined for extraordinary possibilities.


Now, let's visualize the scene.

We will use DALL-E to generate an image.

In [ ]:
response = client.images.generate(
    model="dall-e-3",
    prompt=story,
    n=1,
    size="1024x1024"
)

image = response.data[0].url

print(image)

https://oaidalleapiprodscus.blob.core.windows.net/private/org-vezYUtUuPSSYHgIxYnniZxE4/user-auf1V9TNHptTVhs0XzyQi2Iq/img-Qoeu1S1SVkHirkvFz9Yw5UE7.png?st=2025-05-13T13%3A34%3A04Z&se=2025-05-13T15%3A34%3A04Z&sp=r&sv=2024-08-04&sr=b&rscd=inline&rsct=image/png&skoid=cc612491-d948-4d2e-9821-2683df3719f5&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2025-05-13T06%3A53%3A15Z&ske=2025-05-14T06%3A53%3A15Z&sks=b&skv=2024-08-04&sig=3p0maegWQlWl0F5dVIy15IoXFGiFTRHp8aYL2E78xhE%3D


Now, let's narrate the story.

We will use the text-to-speech API to convert the story into speech.

In [ ]:
speech_file_path = "./speech.mp3"

with client.audio.speech.with_streaming_response.create(
    model="gpt-4o-mini-tts",
    voice="coral",
    input=story,
    instructions="Speak in a cheerful and positive tone.",
) as response:
    response.stream_to_file(speech_file_path)

Other parameters we get as output in response:

In [ ]:
response = client.chat.completions.create(
    model="gpt-4o",
    messages=[
        {"role": "user", "content": "Hey, how are you?"}
    ]
)

print(response.model_dump_json())

{"id":"chatcmpl-BWl07i3B7IytPUs70HLqr6tNbBRn8","choices":[{"finish_reason":"stop","index":0,"logprobs":null,"message":{"content":"Hello! I'm just a computer program, so I don't have feelings, but I'm here to help you. How can I assist you today?","refusal":null,"role":"assistant","annotations":[],"audio":null,"function_call":null,"tool_calls":null}}],"created":1747146867,"model":"gpt-4o-2024-08-06","object":"chat.completion","service_tier":"default","system_fingerprint":"fp_f5bdcc3276","usage":{"completion_tokens":29,"prompt_tokens":13,"total_tokens":42,"completion_tokens_details":{"accepted_prediction_tokens":0,"audio_tokens":0,"reasoning_tokens":0,"rejected_prediction_tokens":0},"prompt_tokens_details":{"audio_tokens":0,"cached_tokens":0}}}


## Error handling

We will be reproducing 401 and 429 errors.

1. 401 error

In [ ]:
client = OpenAI(api_key="Some Random API Key")

response = client.chat.completions.create(
    model="gpt-4o",
    messages=[
        {"role": "user", "content": "Hello, how are you?"}
    ]
)

print(response.choices[0].message.content)

AuthenticationError: Error code: 401 - {'error': {'message': 'Incorrect API key provided: Some Ran******* Key. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}

2. Now, let us handle the 429 error.

In [ ]:
WRONG_API_KEY = userdata.get("OPENAI_API_KEY_ERROR")

client = OpenAI(api_key=WRONG_API_KEY)

response = client.chat.completions.create(
    model="gpt-4o",
    messages=[
        {"role": "user", "content": "Hello, how are you?"}
    ]
)

print(response.choices[0].message.content)

AuthenticationError: Error code: 401 - {'error': {'message': 'Incorrect API key provided: sk-proj-**********************************************************************************************************************************************************cdef. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}